### PR TITLE
DistributedPlan: replace operation with modLevel

### PR DESCRIPTION
**Pull request by serprex**
Also rename ExtactInsertRangeTableEntry to ExtractResultRelationRTE,
as the source of this function didn't match the documentation

This causes no behaviorial changes, only organizes better to implement modifying CTEs (#2776)

### CODE REVIEW

File: src/backend/distributed/executor/multi_router_executor.c [hunk 1]
```diff
@@ -158,106 +158,106 @@ AcquireMetadataLocks(List *taskList)
...
-		 * A RowExclusiveLock does not conflict with itself and therefore allows
-		 * multiple commutative commands to proceed concurrently. It does
-		 * conflict with ExclusiveLock, which may still be obtained by another
-		 * session that executes an UPDATE/DELETE/UPSERT command with
-		 * citus.all_modifications_commutative disabled.
-		 */
 
-		lockMode = RowExclusiveLock;
-	}
-	else if (task->upsertQuery || commandType == CMD_UPDATE || commandType == CMD_DELETE)
```
Comment (serprex) [inline src/backend/distributed/executor/multi_router_executor.c]: Need to make sure `modLevel == MODLEVEL_NONCOMMUTATIVE` when `task->upsertQuery`

File: src/include/distributed/multi_physical_planner.h [hunk 2]
```diff
@@ -110,6 +110,15 @@ typedef enum
 } BoundaryNodeJobType;
 
 
+/* Enumeration that specifies extent of DML modifications */
+typedef enum ModifyLevel
+{
+	MODLEVEL_NONE = 0,
```
Comment (marcocitus) [inline src/include/distributed/multi_physical_planner.h]: We tend to shun acronyms and shorthands, so maybe MODIFY_LEVEL_.

File: src/backend/distributed/executor/multi_router_executor.c [hunk 3]
```diff
@@ -158,106 +158,106 @@ AcquireMetadataLocks(List *taskList)
...
  */
 static void
-AcquireExecutorShardLock(Task *task, CmdType commandType)
+AcquireExecutorShardLock(Task *task, ModifyLevel modLevel)
 {
-	LOCKMODE lockMode = NoLock;
 	int64 shardId = task->anchorShardId;
 
-	if (commandType == CMD_SELECT)
+	if (shardId != INVALID_SHARD_ID)
```
Comment (marcocitus) [inline src/backend/distributed/executor/multi_router_executor.c]: Could we do an early return if `shardId == INVALID_SHARD_ID`? Long code blocks are hard to follow and this makes it seem like we may have a task without a shard ID that does have a relevant `relationRowLockList` or for which `RequiresConsistentSnapshot` is true, which I don't think can be the case.

File: src/backend/distributed/executor/multi_router_executor.c (hunk 1 above)
Comment (marcocitus) [reply src/backend/distributed/executor/multi_router_executor.c]: Hmz, seems like we're missing some test coverage.

File: src/backend/distributed/master/master_truncate.c [hunk 4]
```diff
@@ -79,7 +79,7 @@ citus_truncate_trigger(PG_FUNCTION_ARGS)
 
 		if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
 		{
-			ExecuteModifyTasksSequentiallyWithoutResults(taskList, CMD_UTILITY);
+			ExecuteModifyTasksSequentiallyWithoutResults(taskList, MODLEVEL_NONE);
```
Comment (marcocitus) [inline src/backend/distributed/master/master_truncate.c]: This reads a little strangely, since we're clearly doing a modification here. I wonder whether we should have a MODIFY_LEVEL_DDL to indicate that this is a non-commutative modification, but we don't need special locks because it takes an AccessExclusiveLock on the table.

File: src/backend/distributed/executor/multi_router_executor.c (hunk 3 above)
Comment (marcocitus) [reply src/backend/distributed/executor/multi_router_executor.c]: Discussed this offline, it seems we may want to leverage `relationRowLockList` for CTEs.

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 5]
```diff
@@ -4310,6 +4309,41 @@ GenerateSyntheticShardIntervalArray(int partitionCount)
...
+	switch (query->commandType)
+	{
+		case CMD_SELECT:
+		{
+			return ROW_MODIFY_READONLY;
+		}
+
+		case CMD_INSERT:
+		{
+			if (query->onConflict != NULL)
```
Comment (pykello) [inline src/backend/distributed/planner/multi_physical_planner.c]: I'm a bit confused here. Why is an `INSERT ... ON CONFLICT ...` is commutative? Am I misunderstanding this condition?

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 6]
```diff
@@ -4310,6 +4309,41 @@ GenerateSyntheticShardIntervalArray(int partitionCount)
...
+			{
+				return ROW_MODIFY_COMMUTATIVE;
+			}
```
Comment (pykello) [inline src/backend/distributed/planner/multi_physical_planner.c]: Minor: without an `else` and with the `FALLTHROUGH` I need to scan more lines to understand this. I'd have preferred an `else` block here.

File: src/backend/distributed/planner/multi_physical_planner.c (hunk 5 above)
Comment (serprex) [reply src/backend/distributed/planner/multi_physical_planner.c]: Right, originally the condition was `== NULL` & then in a daze looking into why things weren't working I read it thinking "That's not right" & flipped to `!= NULL`

File: src/backend/distributed/planner/distributed_planner.c [hunk 7]
```diff
@@ -457,19 +457,12 @@ IsMultiTaskPlan(DistributedPlan *distributedPlan)
...
 {
-	CmdType commandType = distributedPlan->operation;
-
-	if (commandType == CMD_UPDATE || commandType == CMD_DELETE)
-	{
-		return true;
-	}
-
-	return false;
+	return distributedPlan->modLevel >= ROW_MODIFY_NONCOMMUTATIVE;
```
Comment (marcocitus) [inline src/backend/distributed/planner/distributed_planner.c]: I think the only caller of this function is `IsMultiShardModifyPlan`, which is quite specifically looking for multi-shard UPDATE/DELETE and not multi-shard upserts. It might be ok to continue checking command type for now, at least until we can push down CTEs in multi-shard statements.

For context: `IsMultiShardModifyPlan` function is called in `CreateDistributedPlannedStmt` to avoid planning prepared update/delete command that has a `distribution column = $1` filter as a multi-shard update/delete when it could be a single-shard update/delete if the parameter was resolved. For multi-shard upserts we can tolerate parameters in the distribution column since we can defer shard pruning to the executor, so you could see a slight performance improvement when using a prepared statement.

File: src/backend/distributed/executor/adaptive_executor.c [hunk 8]
```diff
@@ -968,16 +968,11 @@ DistributedExecutionRequiresRollback(DistributedExecution *execution)
  * that contains FOR UPDATE clause that locks any reference tables.
  */
 static bool
-SelectForUpdateOnReferenceTable(CmdType operation, List *taskList)
+SelectForUpdateOnReferenceTable(RowModifyLevel modLevel, List *taskList)
 {
 	Task *task = NULL;
 	ListCell *rtiLockCell = NULL;
 
-	if (operation != CMD_SELECT)
```
[the remark this answers is not in the document]
Comment (serprex) [reply src/backend/distributed/executor/adaptive_executor.c]: I'd like to change `SELECT FOR UPDATE` to be `ROW_MODIFY_NONCOMMUTATIVE` but don't currently know enough to know if that'd make sense in the end

File: src/backend/distributed/executor/adaptive_executor.c [hunk 9]
```diff
@@ -1402,7 +1401,7 @@ ExecutionOrderForTask(CmdType operation, Task *task)
 
 		case MODIFY_TASK:
 		{
-			if (operation == CMD_INSERT && !task->upsertQuery)
+			if (modLevel >= ROW_MODIFY_NONCOMMUTATIVE)
```
[the remark this answers is not in the document]
Comment (serprex) [reply src/backend/distributed/executor/adaptive_executor.c]: Reading over it, this didn't make sense at first. Found comment at top explaining that the weaker locks made by inserts open them up to deadlock, so it makes sense to change this to `< ROW_MODIFY_COMMUTATIVE` as opposed to `== ROW_MODIFY_COMMUTATIVE`. Just wanting to make sure the mental model of what `RowModifyLevel` isn't breaking down here

**Comment by serprex** [comment]
Squashed. This is ready to merge now
